### PR TITLE
esp8266/modules/webrepl: Simplify the WebREPL setup

### PR DESCRIPTION
**Pull request by deshipu**
Instead of a poor man's user interface with menus, simply let the
user specify the password on the command line, and provide two commands,
one for enabling, and one for disabling the startup.

This patch also makes the line matching much more strict, so that a
comment like "# Here we import webrepl" doesn't confuse the machinery
and lead to a syntax error. Only lines that look exactly like the line
we generate are ever modified.

If there are multiple matching lines, only the first one is left, the
rest are deleted.

### CODE REVIEW

**Comment by deshipu** [comment]
This is my proposal for the webrepl setup. I didn't include documentation changes, but I will add that if there is positive feedback for it.

Note that even if this is rejected, it would still be good to fix the error with the comment somehow.

**Comment by deshipu** [comment]
On my machine, this change makes the resulting binary 1772 bytes smaller.

**Comment by deshipu** [comment]
This tricky automatic processing of user-editable code got me thinking... maybe it would be better, to have a separate module, like `webrepl_setup.py` or similar, that is *always* getting imported in the `boot.py`, but that should contain no user code. Then we can just happily rewrite the whole file, and simplify things even more.

**Comment by pfalcon** [comment]
Sorry, but webrepl setup tries to replicate that of a typical Unix (Linux) service. For example, in Ubuntu, to start a service, you do:

~~~
service ssh start
~~~

to stop

~~~
service ssh stop
~~~

Then some setup for running a service on boot, and other management operations.

Nowhere you pass password or other configuration in the above scheme, it's stored and managed separately. `webrepl.start()`, `webrepl.stop()` directly implement operations like above, and `webrepl_setup` implements "user-friendly" configuration, as there's no editor to let users edit config files.

`webrepl.start()` accepts password only by the request and as the courtesy to 512KB module users who don't have a filesystem to store config files.

Summing up: the aim is not to do it with minimal code, the aim is to follow well known patterns (best-practices), and make it easy to unsuspecting users (those who don't know that they should edit boot.py, and have no idea how to do that).

File: esp8266/modules/webrepl.py [hunk 1]
```diff
@@ -75,3 +70,51 @@ def start_foreground(port=8266):
...
+    # Only allow printable ASCII characters.
+    for c in password:
+        if not 32 <= ord(c) <= 126:
+            raise ValueError("Invalid character in the password")
+    # Escape backslashes and quotes.
+    password = password.replace('\\', '\\\\').replace("'", "\\'")
+    return password
+
+def _change_daemon(enable, password=''):
+    LINE_START = "import webrepl; webrepl.start(password='"
```
Comment (pfalcon) [inline esp8266/modules/webrepl.py]: Let me return your question: What if a user splits that into 2 lines? What if one puts 2 spaced after ";"? 3? 5? what if they use `"` as a string delimiter? I hope it's now clear that it's a dumb question. "If a user wants to fool the script, they can - but they fool themselves, not script." However, you really try to "parse the code", while I tried to minimize need for that as much as possible. Thus, in real-world conditions your script would fail with a higher probability, while violating the config-vs-code separation best practice.

Comment (deshipu) [inline esp8266/modules/webrepl.py]: It won't fail. It won't touch any modified user code, which is by design. User data (and code) is sacred.

If the user is advanced enough to modify those lines themselves, they are advanced enough to not rely on this script.

Also, see my note in the comment about moving this whole thing into a separate python module, as a saner solution to this problem.

**Comment by pfalcon** [comment]
> This tricky automatic processing of user-editable code got me thinking...

Thinking is always good. Oftentimes, you need to actually start doing something to get good thoughts about it. Just imagine that to change MySQL password you'd need to dig deep into its code, and change line no.374 in file foo/bar/baz/stuff.c. Then to recompile it and re-install. Same story with Postgres. And to change your account password, you'd need to recompile the kernel and entire userspace, just in case it doesn't stick somewhere.

Don't do it like that, follow the best practices, separate config from code.

**Comment by deshipu** [comment]
I'm not sure if making just this one thing behave completely different than the whole rest of MicroPython is indeed a good way of avoiding confusion. It also makes it practically impossible to do this from code.

If you really care about not surprising users, then maybe we can use those 1772 bytes we have gained here to add that `errno` attribute to exceptions, make a constructor for the slice object, add support for negative step slices, and a bunch of other differences from standard Python?

File: esp8266/modules/webrepl.py (hunk 1 above)
Comment (pfalcon) [inline esp8266/modules/webrepl.py]: > It won't fail. It won't touch any modified user code, which is by design.

Then happens what? webrepl gets started 2 times?

**Comment by deshipu** [comment]
> Just imagine that to change MySQL password you'd need to dig deep into its code, and change line no.374 in file foo/bar/baz/stuff.c. Then to recompile it and re-install.

I will just point out in this spot that you are already saving the user password in a Python code file, and already overwriting that file from the webrepl code, and that, yes, editing that file is exactly the way you suggested for changing the password manually. Not cool, man.

**Comment by pfalcon** [comment]
> those 1772 bytes

Not all bytes are created equal, that's what you seem to miss. We grudge about bytes in uPy core, because every and each user will be affected by them. But those 1772 bytes is a user-level application. The best way to save bytes on user application is to exclude it completely (and that should be easy for any user, and we even ship only in release builds, someone building from master wouldn't get it at all). However, while it's included, and is an always-running service, it's better to follow known best practices for running services.

**Comment by deshipu** [comment]
Let's compare the user experiences of your and mine approaches, shall we?

Right now, to enable WebREPL, an user has to:

- Find in the documentation the instructions for doing so, to know what to import.
- Connect to the REPL over serial, and type `import webrepl_setup`
- Read a screenful of English text (good if he's a native speaker) which uses words and terms that are unclear and quite domain-specific.
- Select the correct options in correct order (wrong order will result in an error).
- Type the password twice without any feedback, possibly with CapsLock on or some similar silly mistake.
- Possibly repeat that if the password was rejected.
- Select more options, answer the question about restarting.

With what I propose, the user has to:

- Find in the documentation the instructions for doing so, to know what to do.
- Import `webrepl`
- Run `webrepl.enable("your password")`
- Possibly repeat with a different password if the password was rejected.
- That's it.

**Comment by pfalcon** [comment]
> I will just point out in this spot that you are already saving the user password in a Python code file, and already overwriting that file from the webrepl code, and that, yes, editing that file is exactly the way you suggested for changing the password manually. Not cool, man.

You didn't see what they do with Ruby, man. I reuse the fact that simple config file format can be parsed as a Python code. Why bother to write a parser, even simple, if Python parser is already written and always there, so takes 0 bytes? So, `webrepl_cfg.py` is a `*separate*` `*config*` file. It's not code from user's perspective.

**Comment by deshipu** [comment]
> Not all bytes are created equal, that's what you seem to miss. We grudge about bytes in uPy core, because every and each user will be affected by them. But those 1772 bytes is a user-level application.

It is a user-level application that is getting included in the compiled image by default. Which makes it no different from a feature that the user can disable in the makefile before compilation.

**Comment by pfalcon** [comment]
> * Find in the documentation the instructions for doing so, to know what to import.
> * Connect to the REPL over serial, and type import webrepl_setup

That's all what's needed in my approach, thru everything else user is guided by the hand.

Anyway: I tried to make you understand why it was done like it was done. I don't know if you tried to understand or not. There's not a single way to do it, and our solutions are at some level are the same. We aren't changing anything for the upcoming release, we're in freeze on that. Afterwards, if you collect enough votes, your solution *could* be merged. But it all boils down to maintenance. Are you going to maintain it? Probably not, and it's a core part of the project, at least for now (until we declare KS project to be "done"). I won't maintain your script (I wouldn't maintain even mine, but I have to, as a maintainer).

**Comment by deshipu** [comment]
I added a commit that makes it use a separate python module for all script-editable content, and thus will never mangle user code in `boot.py`.

**Comment by deshipu** [comment]
@pfalcon this is not  a popularity contest. Let's just let the emotions cool down, and then pick the solution that is technically better, has fewer opportunities for something to go horribly wrong and is easier to maintain.

**Comment by pfalcon** [comment]
> technically better

Technical stuff is not a top-level goal. Paradigmatical requirements are. And organizational, like who maintains it. Argue your point to @dpgeorge, maybe he'll want to be a backup maintainer for your code. From my POV, it's not worth time - there're real things to work on, like HW API.

Oh, btw - don't put anything into existing webrepl.py. You're trying to deprive 512KB module users from running it (or their apps together with it).

**Comment by deshipu** [comment]
Can you at least consider the trick from the last commit, so that the setup doesn't horribly mangle every line of boot.py that contains the word "webrepl"? The status checking also becomes easier then -- you just check if the file is empty.

**Comment by deshipu** [comment]
There is one more thing to consider -- 3rd party tools, similar to "ampy", the "mu" editor or various GUI programs for flashing and configuring the board. They might want to implement an "enable WebREPL" checkbox or option. With how it is now, it's very hard to call the `webrepl_setup` functions programmatically -- so they will all end up re-implementing the logic, with their own set of bugs and security holes. Just something to think about.

**Comment by pfalcon** [comment]
> Can you at least consider the trick from the last commit

Not for this release. Code freeze. I don't have time to retest this stuff.

And what trick? Adding extra level of module import? No, it's wasteful.

> There is one more thing to consider -- 3rd party tools, similar to "ampy", the "mu" editor or various GUI programs for flashing and configuring the board. They might want to implement an "enable WebREPL" checkbox or option. With how it is now, it's very hard to call the webrepl_setup

They can. By adding or removing comments (and only comments) at 2 lines of boot.py. They also can change password on user's behalf - by changing webrepl_cfg.py. webrepl_setup is a user-level application, not an interface.


I told you that you have rather narrow view. You're trying to slide thru problem space with a microscope, consider one side of a problem, then another. When I wrote that code, I kept in mind 5-6 requirements, and the code adheres to them. Is it perfect? Hell, no. But it works and does what it's designed to. Can it be improved? Possibly, but why? What are the requirements for that? Are they compatible with the original requirements? Is it worth to answer these questions? In the end, if *you* don't like what I wrote, you don't have to use it, it's a user app.

Please read some stuff on the problem above ("narrow view"), e.g. http://netundocumented.com/blog/2005/09/programmer-myopia . And generally, stick to your area of core expertise and slowly work on the areas you're not familiar with - it takes time. I give that suggestion hoping that it really may help you get better perspective at how things can be done differently.

Otherwise, I don't have time any more to discuss this and invoke #2602.

**Comment by deshipu** [comment]
>> Not all bytes are created equal, that's what you seem to miss. We grudge about bytes in uPy core, because every and each user will be affected by them. But those 1772 bytes is a user-level application.

> It is a user-level application that is getting included in the compiled image by default. Which makes it no different from a feature that the user can disable in the makefile before compilation.

I'm very sorry, I was actually wrong here. The user-level application is very different from the standard Python features that are omitted, in that a user can always add any additional code they desire, without having to maintain their own fork of the whole project.

> Technical stuff is not a top-level goal.

What happened to meritocracy?

> Paradigmatical requirements are.

That's not even a word. Can you point me to where those requirements for the webrepl setup are defined?

> Sorry, but webrepl setup tries to replicate that of a typical Unix (Linux) service. For example, in Ubuntu, to start a service, you do: [...]
> Summing up: the aim is not to do it with minimal code, the aim is to follow well known patterns (best-practices), and make it easy to unsuspecting users (those who don't know that they should edit boot.py, and have no idea how to do that).

This is called "cargo cult" -- trying to reproduce unimportant details of a solution used in a completely different context, without understanding how it actually works and what are the effects of it.

> That's all what's needed in my approach, thru everything else user is guided by the hand.

You are making three assumptions that are not always true here. First, you assume that the user will read the messages you display. Second, you assume the user will be able to understand the messages. Third, you assume the user will be able to follow the instructions. Those assumptions make this solution only work in a narrow set of circumstances -- when the user has full connection over serial to the board (and not, for instance, just tx) and a screen large enough to display the whole of the text, when the user can read and understand English fluently, including the domain-specific technical terms (and not, for instance, when following a tutorial translated to a different language), when the user has a full keyboard and can type individual characters (and not, for instance, copy-pastes the commands) and actually understands how this type of menus works.

> And what trick? Adding extra level of module import? No, it's wasteful.

Should I point again that you are already importing that module?

> And generally, stick to your area of core expertise and slowly work on the areas you're not familiar with - it takes time.

It so happens that I used to freelance a couple of years in the area of user experience and accessibility design. This design has such glaring problems, that it doesn't take an expert to see it anyways. You just have to step away for a moment and get over your programmer myopia.

> Otherwise, I don't have time any more to discuss this and invoke #2602.

How convenient. Shame you didn't invoke it when you invited us on the forum today to write this code.
